lib/vrl/stdlib/src/tag_types_externally.rs

```diff
@@ -0,0 +1,253 @@
+use shared::btreemap;
+use std::collections::BTreeMap;
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct TagTypesExternally;
+
+impl Function for TagTypesExternally {
+    fn identifier(&self) -> &'static str {
+        "tag_types_externally"
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "object",
+            source: indoc! {r#"
+                tag_types_externally({
+                    "message": "Hello world",
+                    "request": {
+                        "duration_ms": 67.9
+                    }
+                })
+            "#},
+            result: Ok(
+                r#"{ "message": { "bytes": "Hello world" }, "request": { "duration_ms": { "float": 67.9 } } }"#,
+            ),
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+
+        Ok(Box::new(TagTypesExternallyFn { value }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            kind: kind::ANY,
+            required: true,
+        }]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct TagTypesExternallyFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for TagTypesExternallyFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let tagged_externally = tag_type_externally(&value);
+
+        Ok(tagged_externally)
+    }
+
+    fn type_def(&self, _: &state::Compiler) -> TypeDef {
+        TypeDef::new()
+            .infallible()
+            .object::<(), Kind>(map! {
+                (): Kind::all()
+            })
+            .add_array_mapped::<(), Kind>(map! {
+                (): Kind::all()
+            })
+    }
+}
+
+fn tag_type_externally(value: &Value) -> Value {
```

Comment:
Couldn't this take `Value` by value instead of reference?

Reply:
Yes, I was also unsure about this. Why would we want this to be consuming when it doesn't need to be, though?

Reply:
Well, it _does_ need to be, since you're now cloning the values inside?

But I haven't looked too deep, so there might be a reason somewhere that prevents you from taking ownership here.

Reply:
So what I'm proposing here is:

1. take by value
2. remove all the `clone()` calls for scalar values
3. update the collection value's to use `into_iter`

Reply:
I confused myself there, thought that consuming the value could mutate nested values, which obviously isn't the case. Thanks for the suggestion!